api/unstable/fingerprint.h

```diff
@@ -35,6 +35,102 @@ typedef enum {
+ * @returns S2N_SUCCESS on success, S2N_FAILURE on failure.
+ */
+S2N_API int s2n_fingerprint_reset(struct s2n_fingerprint *fingerprint);
+
+/**
+ * Sets the ClientHello to be fingerprinted.
+ *
+ * @param fingerprint The fingerprint to be modified
+ * @param ch The client hello to be fingerprinted. It will not be copied, so needs
+ * to live at least as long as this fingerprinting operation.
+ * @returns S2N_SUCCESS on success, S2N_FAILURE on failure.
+ */
+S2N_API int s2n_fingerprint_set_client_hello(struct s2n_fingerprint *fingerprint, struct s2n_client_hello *ch);
+
+/**
+ * Get the size of the fingerprint hash.
+ *
+ * Fingerprint hashes should be a constant size, but that size will vary based
+ * on the fingerprinting method used.
+ *
+ * @param fingerprint The fingerprint to be used for the hash
+ * @param size Output variable to be set to the size of the hash
+ * @returns S2N_SUCCESS on success, S2N_FAILURE on failure.
+ */
+S2N_API int s2n_fingerprint_get_hash_size(const struct s2n_fingerprint *fingerprint, uint32_t *size);
+
+/**
+ * Calculates a fingerprint hash.
+ *
+ * @param fingerprint The fingerprint to be used for the hash
+ * @param max_output_size The maximum size of data that may be written to `output`.
+ * If `output` is too small, an S2N_ERR_T_USAGE error will occur.
+ * @param output The location that the requested hash will be written to.
+ * @param output_size Output variable to be set to the actual size of the data
+ * written to `output`.
+ * @returns S2N_SUCCESS on success, S2N_FAILURE on failure.
+ */
+S2N_API int s2n_fingerprint_get_hash(struct s2n_fingerprint *fingerprint,
+        uint32_t max_output_size, uint8_t *output, uint32_t *output_size);
+
+/**
+ * Get the size of the raw fingerprint string.
+ *
+ * The size of the raw string depends on the ClientHello and cannot be known
+ * without calculating the fingerprint. Either `s2n_fingerprint_get_hash` or
+ * `s2n_fingerprint_get_raw` must be called before this method.
+ *
+ * @param fingerprint The fingerprint to be used for the raw string
+ * @param size Output variable to be set to the size of the raw string
+ * @returns S2N_SUCCESS on success, S2N_FAILURE on failure.
+ */
+S2N_API int s2n_fingerprint_get_raw_size(const struct s2n_fingerprint *fingerprint, uint32_t *size);
+
+/**
+ * Calculates the raw string for a fingerprint.
+ *
+ * @param fingerprint The fingerprint to be used for the raw string
+ * @param max_output_size The maximum size of data that may be written to `output`.
+ * If `output` is too small, an S2N_ERR_T_USAGE error will occur.
+ * @param output The location that the requested hash will be written to.
+ * @param output_size Output variable to be set to the actual size of the data
+ * written to `output`.
+ * @returns S2N_SUCCESS on success, S2N_FAILURE on failure.
+ */
+S2N_API int s2n_fingerprint_get_raw(struct s2n_fingerprint *fingerprint,
+        uint32_t max_output_size, uint8_t *output, uint32_t *output_size);
```

Comment:
I've gone back and forth on including or not including the raw / full string. Currently no customers are using it, but customers did originally ask for it. With the s2n_fingerprint_hash struct I don't think calculating it adds much complexity, but it does still add some complexity. If we want to remove the flexibility, we also have to remove the old version of the method, but again this API is marked unstable.

Reply:
It seems very possible that people might have some new usecase for this in the future, so I like having it included now to ensure that it always fits in well with the feature.